### PR TITLE
added modal for uploading files

### PR DESCRIPTION
**Pull request by badescuga**
When you start drag-and-dropping files, you see the percent near the loader as till now. 
screenshot_12

Once you push the percentage button, you get a modal showing your recent uploaded files, and the ones in progress. 

screenshot_13

**Performance impact** (rated in according with last week's pull request).

When the modal is closed, the impact is 0. Extra, i added an auto-cleaner for the cache once it got too big.

When the modal is opened, the only extra impact is the draw of the knockout collection object.

=================

**Personal observations**
The modal performed quite well even with hundred of files. Didn't see any lag or drawbacks. Did also clean the previous implementation a bit.

### CODE REVIEW

**Comment by shrimpy** [comment]
image

Love this feature, it is good when uploading large files and see the progress.
Notice that the percentage is for each file. if i drap and drop multiple files, i will see multiple "100%". will be good to show an overall progress, e.g drap and drop 10 files, once a file is uploaded, progress 10%.

File: Kudu.Services.Web/Content/Scripts/FileBrowser.js [hunk 1]
```diff
@@ -80,39 +80,47 @@ function showAceHelpModal() {
...
+        this._infoMessage = message;
     },
-    addCopyStats: function (uri, loadedData, totalData
-   )  { 
-        if (copyProgressObjects[uri]) {
-            if (loadedData === totalData) { 
-                copyProgressObjects[uri].endDate = $.now();
+    addCopyStats: function (uri, loadedData, totalData) {
+
+        uri = uri.substring(uri.indexOf('/vfs')+5, uri.length); // slice uri to be prettier
```
Comment (shrimpy) [inline Kudu.Services.Web/Content/Scripts/FileBrowser.js]: for "uri", will be good to add a comment to have a sample uri. make reading the code easier.

File: Kudu.Services.Web/Content/Scripts/FileBrowser.js [hunk 2]
```diff
@@ -80,39 +80,47 @@ function showAceHelpModal() {
...
+                this._copyProgressObjects[uri].copyPackEnded = false;
             }
         } else {
-            copyProgressObjects[uri] = {};
-            copyProgressObjects[uri].startDate = $.now();
-            //this is used for when copying multiple files in the same time so that i may stii have a coherent percentage
-            copyProgressObjects[uri].transactionPackFinished = false; 
+            this._copyProgressObjects[uri] = {};
+            this._copyProgressObjects[uri].startDate = $.now();
+            this._copyProgressObjects[uri].copyPackEnded = false; //this is used for when copying multiple files in the same time so that i may stii have a coherent percentage
```
Comment (shrimpy) [inline Kudu.Services.Web/Content/Scripts/FileBrowser.js]: typo "still" ?

File: Kudu.Services.Web/Content/Scripts/FileBrowser.js [hunk 3]
```diff
@@ -503,9 +535,43 @@ $.connection.hub.start().done(function () {
...
+        
+            if (perc == 100 || perc == 0) {
+                viewModel.isTransferInProgress(false);
+            } else {
+                viewModel.isTransferInProgress(true);
+            }
+
+            //handler for clearing out cache once it gets too large 
+            var currentObjCount = Object.keys(copyObjs).length;
+           // console.log('CURRENT OBJ COUNT: %s', currentObjCount);
```
Comment (shrimpy) [inline Kudu.Services.Web/Content/Scripts/FileBrowser.js]: remove unused debug code

**Comment by badescuga** [comment]
Great to hear you like it! The percentage is done for all files that are
being copied. You see the percentage restarting because i start counting
for files that start uploading. Not all files start as once(some have to
finish uploading before others start).

You can see multifile progress by copying 2 larger files.

**Comment by shrimpy** [comment]
RE: You see the percentage restarting because i start counting for files that start uploading. Not all files start as once

I might be picky about the experience, but seeing the progress restarting seems a bit confusing to me.
I would expect the overall progress bar to start from 0 and progress as file uploaded, only reach 100% when all files are done and only show 100% once as it is overall progress.

Another issue for the current implementation is, for the case I try to upload below 8 files. I saw three “100%” and then the overall percentage goes away while file is still uploading in background.
For the overall progress bar, I would suggest to grab the total file count at the beginning, and as a file is uploaded, we progress base on the file count.



[cid:image002.jpg@01D101B7.584031A0]


[cid:image006.jpg@01D101B7.584031A0]

Thanks,
Xiaomin Wu

**Comment by badescuga** [comment]
I agree with you and have thought about implementing at the start, but
tried to keep it as simple as i can. I can redo this.

**Comment by badescuga** [comment]
I'll take a look at this in a couple of hours to see if there is a quick
fix.

**Comment by badescuga** [comment]
Hi guys,

Found another way to do the processing so that I process all the files at start (at least their size). Will cancel the pull request tomorrow (in like 10-12 hours) and will work on a better solution.

Thanks!
From:  Xiaomin Wu
Reply-To:  projectkudu/kudu
Date:  Thursday 8 October 2015 at 20:52
To:  projectkudu/kudu
Cc:  Alex Badescu
Subject:  Re: [kudu] added modal for uploading files (#1729)

RE: You see the percentage restarting because i start counting for files that start uploading. Not all files start as once 

I might be picky about the experience, but seeing the progress restarting seems a bit confusing to me. 
I would expect the overall progress bar to start from 0 and progress as file uploaded, only reach 100% when all files are done and only show 100% once as it is overall progress. 

Another issue for the current implementation is, for the case I try to upload below 8 files. I saw three “100%” and then the overall percentage goes away while file is still uploading in background. 
For the overall progress bar, I would suggest to grab the total file count at the beginning, and as a file is uploaded, we progress base on the file count. 



[cid:image002.jpg@01D101B7.584031A0] 


[cid:image006.jpg@01D101B7.584031A0] 

Thanks, 
Xiaomin Wu

**Pull request closed without merging**